p2p/src/network/pubsub/p2p_network_pubsub_reducer.rs

```diff
@@ -440,6 +472,156 @@ impl P2pNetworkPubsubState {
+                    bug_condition!(
+                        "`P2pNetworkPubsubAction::ValidateIncomingMessage` called on invalid state"
+                    );
+                    return Ok(());
+                };
+
+                let new_message_state = match &content {
+                    GossipNetMessageV2::NewState(block) => {
+                        let block_hash = block.try_hash()?;
+                        P2pNetworkPubsubMessageCacheMessage::PreValidatedBlockMessage {
+                            block_hash,
+                            message,
+                            peer_id,
+                            time,
+                        }
+                    }
+                    _ => P2pNetworkPubsubMessageCacheMessage::PreValidated {
+                        message,
+                        peer_id,
+                        time,
+                    },
+                };
+                pubsub_state
+                    .mcache
+                    .map
+                    .insert(message_id, new_message_state);
+
+                let dispatcher = state_context.into_dispatcher();
+
+                match content {
+                    GossipNetMessageV2::NewState(block) => {
+                        let best_tip = BlockWithHash::try_new(block.clone())?;
+                        dispatcher.push(P2pPeerAction::BestTipUpdate { peer_id, best_tip });
+                        return Ok(());
+                    }
+                    GossipNetMessageV2::TransactionPoolDiff { message, nonce } => {
+                        let nonce = nonce.as_u32();
+                        for transaction in message.0 {
+                            dispatcher.push(P2pChannelsTransactionAction::Libp2pReceived {
+                                peer_id,
+                                transaction: Box::new(transaction),
+                                nonce,
+                            });
+                        }
+                    }
+                    GossipNetMessageV2::SnarkPoolDiff {
+                        message: NetworkPoolSnarkPoolDiffVersionedStableV2::AddSolvedWork(work),
+                        nonce,
+                    } => {
+                        dispatcher.push(P2pChannelsSnarkAction::Libp2pReceived {
+                            peer_id,
+                            snark: Box::new(work.1.into()),
+                            nonce: nonce.as_u32(),
+                        });
+                    }
+                    _ => {}
+                }
+
+                dispatcher.push(P2pNetworkPubsubAction::BroadcastValidatedMessage {
+                    message_id: super::BroadcastMessageId::MessageId { message_id },
+                });
+                Ok(())
+            }
+            P2pNetworkPubsubAction::BroadcastValidatedMessage { message_id } => {
+                let Some((message_id, message)) =
+                    pubsub_state.mcache.get_message_id_and_message(message_id)
+                else {
```

Comment:
same here, is this a bug condition? I assume that by the time these messages are dispatched the message exists. If there are situation in which it will not exist, the enabling condition should probably take care of checking that